src/protocol/templates/ping1d.cpp.in

```diff
@@ -0,0 +1,213 @@
+
+            for (uint16_t i = 0; i < _{{field.name}}_length; i++) {
+                _{{field.name}}[i] = m.{{field.name}}()[i];
+            }
+{% else %}
+            _{{field.name}} = m.{{field.name}}();
+{% endif %}
+{% endfor %}
+        }
+            break;
+
+{% endfor %}
+
+      default:
+          break;
+    }
+}
+
+// ex ping_msg_ping1D_voltage_5 msg(*pd.request(Ping1DNamespace::Voltage_5));
+PingMessage* Ping1D::request(enum Ping1DNamespace::msg_ping1D_id id, uint16_t timeout_ms = 400)
+{
+    ping_msg_ping1D_empty msg;
+    msg.set_id(id);
+    msg.updateChecksum();
+    write(msg.msgData, msg.msgDataLength());
+    return waitMessage(id, timeout_ms);
+}
+
+// ex auto msg = pd.request<ping_msg_ping1D_voltage_5>();
+template <typename T>
+T* Ping1D::request()
+{
+    T resp; // todo there should be some other (static) way to get the message id?
+    static ping_msg_ping1D_empty req;
+    req.set_id(resp.message_id());
+    req.updateChecksum();
+    write(req.msgData, req.msgDataLength());
+    return (T*)waitMessage(resp.message_id());
+}
+
+
+{% for msg in messages["ping1D"]["get"] %}
+bool Ping1D::get_{{msg}}(
+{%- for field in messages["ping1D"]["get"][msg].payload %}
+{% if is_vector(field.type) %}
+{{ "                     " if not loop.first }}{{get_type_string(field.vector.sizetype)}}* {{field.name}}_length, {{get_type_string(field.vector.datatype)}}** {{field.name}}{{ ",\n" if not loop.last }}
+{%- else %}
+{{ "                     " if not loop.first }}{{get_type_string(field.type)}}* {{field.name}}{{ ",\n" if not loop.last }}
+{%- endif %}
+{% endfor -%}
+         )
+{
+
+    if (!request(Ping1DNamespace::{{msg|title}})) {
+        return false; // no reply from device
+    }
+
+    // Copy selected fields
+{% for field in messages["ping1D"]["get"][msg].payload %}
+{% if is_vector(field.type) %}
+    if ({{field.name}}_length) *{{field.name}}_length = _{{field.name}}_length;
+    if ({{field.name}}) *{{field.name}} = _{{field.name}};
+{% else %}
+    if ({{field.name}}) *{{field.name}} = _{{field.name}};
+{% endif %}
+{% endfor %}
+
+    return true; // success
+}
+
+{% endfor %}
+
+{% for msg in messages["ping1D"]["set"] %}
+bool Ping1D::{{msg}}({% for field in messages["ping1D"]["set"][msg].payload %}{{get_type_string(field.type)}} {{field.name}}, {% endfor %}bool verify)
+{
+    ping_msg_ping1D_{{msg}} m;
+{% for field in messages["ping1D"]["set"][msg].payload %}
+    m.set_{{field.name}}({{field.name}});
+{% endfor %}
+    m.updateChecksum();
+    write(m.msgData, m.msgDataLength());
+    if (!request(Ping1DNamespace::{{msg|replace("set_", "")|title}})) {
+        return false; // no reply from device
+    }
+    // Read back the data and check that changes have been applied
+    if (verify
+{% if messages["ping1D"]["set"][msg].payload %}
+        && {% for field in messages["ping1D"]["set"][msg].payload %}_{{field.name}} != {{field.name}}{{ "\n        || " if not loop.last }}{% endfor %}) {
```

Comment:
The `&&` should encompass all `||`;
```
if (verify
        && _scan_start != scan_start
        || _scan_length != scan_length) {
```
Should be:
```
if (verify
        && ( _scan_start != scan_start
        || _scan_length != scan_length)) {
```

Reply:
These two will always evaluate to the same thing, correct?

Reply:
There is () after the &&, the logic result is different. 
![image](https://user-images.githubusercontent.com/1215497/46358921-79e31f80-c63e-11e8-974d-13df3948051a.png)

![image](https://user-images.githubusercontent.com/1215497/46358945-82d3f100-c63e-11e8-82e4-b56bc779f84a.png)

Reply:
You are comparing
`(x ^ y) \/ z` with `x ^ (y \/ z)`

This is not the situation at hand. You should be comparing

`x ^ y \/ z` with `x ^ (y \/ z)`

Reply:
I will test some things here for my own verification and information.

You are right with the correction, I will do it.

Reply:
Yes.
http://cpp.sh/7lmb5